src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -15,325 +15,285 @@
+        return;
     }
-    os_unfair_lock_unlock(&_lock);
-}
 
-- (void)endWork
-{
-    [self.workQueue endWork:self];
-    [self invalidate];
-}
+    // if work item is done (no need to retry), remove from queue and call ready on the next item
+    if (!retry) {
+        [workItem markComplete];
+        [_items removeObjectAtIndex:0];
+    }
 
-- (void)retryWork
-{
-    [self.workQueue retryWork:self];
+    // when "concurrency width" is implemented this will be decremented instead
+    _runningWorkItemCount = 0;
+    [self _callNextReadyWorkItem];
 }
 
-// Called by the work queue
-- (void)callReadyHandlerWithContext:(id)context
+// assume lock is held while calling this
+- (void)_callNextReadyWorkItem
 {
-    dispatch_async(self.queue, ^{
-        os_unfair_lock_lock(&self->_lock);
-        MTRAsyncWorkReadyHandler readyHandler = self->_readyHandler;
-        NSUInteger retryCount = self->_retryCount;
-        if (readyHandler) {
-            self->_retryCount++;
-        }
-        os_unfair_lock_unlock(&self->_lock);
+    // when "concurrency width" is implemented this will be checked against the width
+    if (_runningWorkItemCount) {
+        return; // can't run next work item until the current one is done
+    }
 
-        if (readyHandler == nil) {
-            // Nothing to do here.
-            [self endWork];
-        } else {
-            readyHandler(context, retryCount);
-        }
-    });
-}
+    if (!_items.count) {
+        return; // nothing to run
+    }
 
-// Called by the work queue
-- (void)cancel
-{
-    os_unfair_lock_lock(&self->_lock);
-    dispatch_block_t cancelHandler = self->_cancelHandler;
-    [self _invalidate];
-    os_unfair_lock_unlock(&self->_lock);
-
-    if (cancelHandler) {
-        dispatch_async(self.queue, ^{
-            cancelHandler();
-        });
+    id context = _context;
+    if (!context) {
+        MTR_LOG_ERROR("MTRAsyncWorkQueue context has been lost, dropping queued work items");
+        [_items removeAllObjects];
+        return;
     }
-}
 
-- (void)setBatchingID:(NSUInteger)opaqueBatchingID data:(id)opaqueBatchableData handler:(MTRAsyncWorkBatchingHandler)batchingHandler
-{
-    os_unfair_lock_lock(&self->_lock);
-    _batchable = YES;
-    _batchingID = opaqueBatchingID;
-    _batchableData = opaqueBatchableData;
-    _batchingHandler = batchingHandler;
-    os_unfair_lock_unlock(&self->_lock);
+    // when "concurrency width" is implemented this will be incremented instead
+    _runningWorkItemCount = 1;
+
+    MTRAsyncWorkItem * workItem = _items.firstObject;
+
+    // Check if batching is possible or needed. Only ask work item to batch once for simplicity
+    auto batchingHandler = workItem.batchingHandler;
+    if (batchingHandler && workItem.retryCount == 0) {
+        while (_items.count >= 2) {
+            MTRAsyncWorkItem * nextWorkItem = _items[1];
+            if (!nextWorkItem.batchingHandler || nextWorkItem.batchingID != workItem.batchingID) {
```

Comment:
I think the previous setup sort of considered situations where an item is batchable (so can be merged into previous items) without having a batching handler (i.e. can't have later things merged into it).

Looks like this was not used in practice, so this change to just checking batchingHandler here is probably ok....

Reply:
`setBatchingID:data:handler:` (and `setDuplicateTypeID:handler`) both had non-null handler parameters, so I wasn't aware of this idea. We can re-add that if needed I guess (or maybe it's more obvious in those cases to implement a handler that just doesn't batch anything).

Reply:
I think the new code is fine.